sphinx/environment/__init__.py

```diff
@@ -842,3 +860,208 @@ def _check_toc_parents(toctree_includes: dict[str, list[str]]) -> None:
+        """Return a serial number, e.g. for index entry targets.
+
+        The number is guaranteed to be unique in the current document.
+        """
+        current = self._serial_numbers.get(category, 0)
+        self._serial_numbers[category] = current + 1
+        return current
+
+    # Mapping interface:
+
+    def __getitem__(self, item: str) -> Any:
+        if item == 'annotations':
+            return self.autodoc_annotations
+        if item == 'autodoc:class':
+            return self.autodoc_class
+        if item == 'autodoc:module':
+            return self.autodoc_module
+        if item == 'object':
+            return self.obj_desc_name
+        if item == 'started_at':
+            return self.reading_started_at
+        if item in {
+            '_parser',
+            'autodoc_annotations',
+            'default_domain',
+            'default_role',
+            'docname',
+            'highlight_language',
+            'obj_desc_name',
+            'reading_started_at',
+        }:
+            return getattr(self, item)
+        return self._ext_props[item]
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if key == 'annotations':
+            self.autodoc_annotations = value
+        elif key == 'autodoc:class':
+            self.autodoc_class = value
+        elif key == 'autodoc:module':
+            self.autodoc_module = value
+        elif key == 'object':
+            self.obj_desc_name = value
+        elif key == 'started_at':
+            self.reading_started_at = value
+        elif key in {
+            '_parser',
+            'autodoc_annotations',
+            'default_domain',
+            'default_role',
+            'docname',
+            'highlight_language',
+            'obj_desc_name',
+            'reading_started_at',
+        }:
+            setattr(self, key, value)
+        else:
+            self._ext_props[key] = value
+
+    def __delitem__(self, key: str) -> None:
+        del self._ext_props[key]
+
+    def __contains__(self, item: str) -> bool:
+        if item in {
+            'annotations',
+            'autodoc:class',
+            'autodoc:module',
+            'object',
+            'started_at',
+        }:
+            return True
+        if item in {
+            '_parser',
+            '_serial_numbers',
+            '_ext_props',
+            'autodoc_annotations',
+            'autodoc_class',
+            'autodoc_module',
+            'default_domain',
+            'default_role',
+            'docname',
+            'highlight_language',
+            'obj_desc_name',
+            'reading_started_at',
+        }:
+            return True
+        return item in self._ext_props
+
+    def get(self, key: str, default: Any | None = None) -> Any | None:
+        try:
+            return self[key]
+        except KeyError:
+            return default
+
+    def pop(self, key: str, default: Any | None = None) -> Any | None:
+        if key == 'annotations':
```

Comment:
How about a static dict mapping aliases to normalized keys? (it would reduce the big if duplication. You could also store in this static dict whether the attribute can be popped or not)